src/Language/PureScript/Scheme/CodeGen/Transpiler.hs

```diff
@@ -30,6 +36,38 @@ exprToScheme :: Expr Ann -> AST
 exprToScheme (Literal _ann literal) =
   literalToScheme literal
 
+-- An ADT is translated to a tagged pair. The car of the pair is the tag,
+-- holoding the name of the Constructor as a quoted symbol. The cdr of the
+-- pair is a vector, holding the values of the Constructor, if any.
+--
+-- A Constructor is either a pair literal in the case of a simple sum type
+-- (e.g. data Foo = Bar | Baz) or a a function that produces a tagged pair
+-- holding the values for the constructed ADT.
+--
+-- A product type with N fields results into a curried function of N arguments.
+--     data Foo = Bar Int Int
+--     (lambda (v0) (lambda (v1) (cons 'Bar (vector v0 v1)))) ; defined as Bar
+--
+-- A sum type results into multiple Constructor. If the sum type is `simple'
+-- we just emit a constant representing that constructed value.
+--     data Foo = Bar | Baz
+--     (cons (quote Bar) (vector)) ; defined as Bar
+--     (cons (quote Baz) (vector)) ; defined as Baz
+-- TODO: maybe we could remove that empty vector?
```

Comment:
Yeah you could literally put `nil` in there, since you're never going to access this field anyways

Reply:
True but should we do it in this PR since it's getting pretty big already?

Also: should we do it at all? It's more logic... for what?

Reply:
Addendum: if we want to do this, we might want to just have `'Bar` and `'Baz` instead of `('Bar . #())` or `('Bar . nil)`. It's definitely more lisp-y. But in any case I'd rather do this in a different PR.

Reply:
Sounds good.